File: graph_dfs_problems/01-03_path_one_step3.rb
# グラフのパス (paizaランク C 相当)
# https://paiza.jp/works/mondai/graph_dfs_problems/graph_dfs__path_one_step3

INPUT1 = <<~"EOS"
  3 1 2
EOS
OUTPUT1 = <<~"EOS"
  1 2 3
EOS

INPUT2 = <<~"EOS"
  5 5 3
EOS
OUTPUT2 = <<~"EOS"
  5 4 3 2
EOS

def main(input_str)
  # n: 頂点数, s: 起点, k: 回数
  n, s, k = input_str.split.map(&:to_i)
  # 隣接リスト
  ad_list = Hash.new { [] }
  1.upto(n) do |i|
    1.upto(n) do |j|
      ad_list[i] <<= j if i != j
    end
  end

  # s から k 回移動する経路
  path = [s]
  k.times do
    # 今いる頂点
    cv = path.last
    # 移動可能な未訪問の頂点を末尾に追加する
    ad_list[cv].reverse.each do |nv|
      # 訪問済の頂点はスキップ
      next if path.include?(nv)
      path << nv
      break
    end
  end
  path.join(" ")
end

puts main(INPUT1)
# > 1 3 2
puts main(INPUT2)
# > 5 4 3 2

File: graph_dfs_problems/01-03_path_one_step3.py
# グラフのパス (paizaランク C 相当)
# https://paiza.jp/works/mondai/graph_dfs_problems/graph_dfs__path_one_step3

INPUT1 = """\
3 1 2
"""
OUTPUT1 = """\
1 2 3
"""

INPUT2 = """\
5 5 3
"""
OUTPUT2 = """\
5 4 3 2
"""


def main(input_str):
    # n: 頂点数, s: 起点, k: 回数
    n, s, k = map(int, input_str.split())
    # 隣接リスト
    ad_list = {}
    for i in range(1, n + 1):
        e = [j for j in range(1, n + 1) if j != i]
        ad_list[i] = e

    # s から k 回移動する経路
    path = [s]
    for _ in range(k):
        # 今いる頂点
        cv = path[-1]
        # 移動可能な頂点を選択して walk 末尾へ追加
        for nv in ad_list[cv][::-1]:
            # 訪問済の頂点はスキップ
            if nv in path:
                continue
            # path に追加してループを抜ける
            path.append(nv)
            break

    # 経路を出力
    return " ".join(map(str, path))


print(main(open(0).read()))
